src/librustc/mir/interpret/value.rs

```diff
@@ -237,13 +237,18 @@ impl<'tcx, Tag> Scalar<Tag> {
     }
 
     #[inline]
-    pub fn from_uint(i: impl Into<u128>, size: Size) -> Self {
+    pub fn try_from_uint(i: impl Into<u128>, size: Size) -> InterpResult<'tcx, Self> {
         let i = i.into();
-        assert_eq!(
-            truncate(i, size), i,
-            "Unsigned value {:#x} does not fit in {} bits", i, size.bits()
-        );
-        Scalar::Raw { data: i, size: size.bytes() as u8 }
+        if truncate(i, size) == i {
+            Ok(Scalar::Raw { data: i, size: size.bytes() as u8 })
+        } else {
+            throw_unsup_format!("Unsigned value {:#x} does not fit in {} bits", i, size.bits())
+        }
+    }
+
+    #[inline]
+    pub fn from_uint(i: impl Into<u128>, size: Size) -> Self {
+        Self::try_from_uint(i, size).unwrap()
```

Comment:
In particular,. this is real bad: it allocates a `String` to format the error message, and then deallocates it immediately! That's way more expensive than what we'd like to do here.

Reply:
I never thought about that, ty for the tip :)